Introduce `before_script` which will be executed prior the command

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This adds the possibility to pass gatekeeping commands to the action without actually polluting the real command.

- with https://github.com/laminas/laminas-ci-matrix-action/pull/31, we want to introduce linting of phpunit, psalm, etc. configuration files
- in https://github.com/laminas/laminas-cache/pull/196/files#diff-3fb49ad00d3736f1ca8276f568f5d8bbe8c1604f6187ab5ecf921e1012f47c98R15, we want to install a very specific version of a composer dependency which won't be installed when using `lowest` or `latest` (due to the constraint `^1.0 || ^2.0 || ^3.0`)
- I've added some `xmlstarlet` magic to modify `convertDeprecationsToExceptions` setting within `phpunit.xml.dist` in https://github.com/laminas/laminas-component-installer/blob/3c998d2b2755985c71cd4d5d9b294228f63d3a65/.laminas-ci.json#L6

All these quirks could be passed via that new JOB property instead of passing consecutive commands via the `command` property.

----

I am open to rename the property, I simply did not found a better naming (yet).

boesing committed:
feature: introduce gatekeeper commands which will be executed prior the command

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented:
Is this some sort of `before_script:`? The name "gatekeeper" kinda implies that they stop something 🤔

boesing commented:
They might stop something in case they fail.

We do have that `.laminas-ci/pre_run.sh` which can be created in every project.
Thats kinda similar logic and might stop the command from being executed (depending on how that script is configured).

boesing commented:
So lets take the `xmllint` example from the link in the initial post, in case the linting fails - the command is not being executed.

So instead someone has to pass `xmllint ... && whateverUsesTheLintedXmlConfig`, we just have an additional entry in that new introduced property.

Ocramius commented:
> We do have that `.laminas-ci/pre_run.sh` which can be created in every project.
 > Thats kinda similar logic and might stop the command from being executed (depending on how that script is configured).

I don't fully understand how this new syntax is different, and whether it's supposed to replace the old one (to be deprecated, I suppose?) :thinking:

boesing commented:
> I don't fully understand how this new syntax is different, and whether it's supposed to replace the old one (to be deprecated, I suppose?) 🤔

What syntax? Not sure if we are talking about the same thing.

1. we still support `pre-run.sh` since that is a script which can be created by projects using our CI action, no plans to drop that
2. we/I want to add a new way to provide additional commands to be executed via the matrix action

So the latter would be stuff like:

1. linting of tool configurations (such as xmllint for psalm.xml/psalm.xml.dist/phpunit.xml/phpunit.xml.dist and yamllint for whatever YAM configuration, etc.) - see https://github.com/laminas/laminas-ci-matrix-action/pull/31
2. whatever else could be added as a command which needs to run beforehand which we are aware of in the matrix already

---

The current way of doing so would be polluting the JOB JSON by doing:

```json
{"...": "...", "command": "xmllint psalm.xml.dist && vendor/bin/psalm --shepherd --no-cache --whatever"}
```

IMHO, this is not really readable and thus I would prefer something like:

```json
{"...": "...", "command": "vendor/bin/psalm --shepherd --no-cache --whatever", "gatekeeper_commands": ["xmllint psalm.xml.dist"]}
```
__(Still open to rename the property tho)__

Ocramius commented:
So, if I understand it correctly, `gatekeeper_commands` (to be renamed, indeed) is a better way of writing `bash -c "$GATEKEEPER_COMMAND" && bash -c "$CI_COMMAND"`?

Looking good, just probably better to call it `before_script` (like in https://docs.gitlab.com/ee/ci/yaml/#before_script)

boesing commented:
> [...]  is a better way of writing bash -c "$GATEKEEPER_COMMAND" && bash -c "$CI_COMMAND"?

Exactly. Not sure if this is really necessary tho but given the fact that we do have some stuff depending on the commands name (BC compatibility for exclusion logic), adding additional stuff to the `command` property might become a burden when it comes to that exclusion logic.

> call it `before_script`

Sounds good, will adjust the code when I find some time for it. Thanks!

boesing committed:
docs: add missing PHP 8.1 to the list of supported PHP versions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: rename `gatekeeper_commands` to `before_script`

Gitlab uses `before_script` as well and therefore, using this to keep some kind of familiar naming seems legit.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: add a hint to the documentation that the `pre-run.sh` can be omitted when `before_script` is being passed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
I've renamed `gatekeeper_commands` to `before_script` and extended the README to outline the idea of `before_script`.

I also created #104 to have an additional `after_script`. I am thinking about moving some quirks to the matrix action after we released a container version which supports `before_script` and `after_script`:

https://github.com/laminas/laminas-continuous-integration-action/blob/c2d713a98f34102e6b46cf31ebd3c92dca9245b9/entrypoint.sh#L195-L199


https://github.com/laminas/laminas-continuous-integration-action/blob/c2d713a98f34102e6b46cf31ebd3c92dca9245b9/entrypoint.sh#L201-L209

Ocramius reviewed (approved):
LGTM :+1: 

Thanks @boesing
